File: test/src/compiler/cases/cpp_integration/aggregate.ts
import { Point, Aggregate, LargerAggregate } from "./declarations/cpp"

const point = new Point(1, 1);

const array = ["this", "is", "string", "array"];
const s = "this is string";
const n = 1;

const aggregate = new Aggregate(point, array, s, n);

console.assert(aggregate.getPoint().x() === 1, "Aggregate.getPoint.x failed");
console.assert(aggregate.getPoint().y() === 1, "Aggregate.getPoint.y failed");

console.assert(aggregate.getString() === "this is string", "Aggregate.getString failed");
console.assert(aggregate.getNumber() === 1, "Aggregate.getNumber failed");

const stringArrayFromAggregate = aggregate.getStringArray();
for (let i = 0; i < stringArrayFromAggregate.length; ++i) {
    console.assert(stringArrayFromAggregate[i] === array[i], "Aggregate.getStringArray failed");
}

const m = new LargerAggregate(1, 1, 11, 11);
const r = m.getRect();
console.assert(r.getSquare() === 100, "Square test failed");

const topLeft = m.getTopLeft();
console.assert(topLeft.x() === 1, "Top left x test failed");
console.assert(topLeft.y() === 1, "Top left y test failed");

const bottomRight = m.getBottomRight();
console.assert(bottomRight.x() === 11, "Bottom right x test failed");
console.assert(bottomRight.y() === 11, "Bottom right y test failed");

const scaled = m.getScaled(10);
console.assert(scaled.getRect().getSquare() === 10000, "Scaled square test failed");

const scaledTopLeft = scaled.getTopLeft();
console.assert(scaledTopLeft.x() === 10, "Scaled top left x test failed");
console.assert(scaledTopLeft.y() === 10, "Scaled top left y test failed");

const scaledBottomRight = scaled.getBottomRight();
console.assert(scaledBottomRight.x() === 110, "Scaled bottom right x test failed");
console.assert(scaledBottomRight.y() === 110, "Scaled bottom right y test failed");

File: test/src/compiler/cases/cpp_integration/cpp/aggregate.cpp
/*
 * Copyright (c) New Cloud Technologies, Ltd., 2014-2022
 *
 * You can not use the contents of the file in any way without
 * New Cloud Technologies, Ltd. written permission.
 *
 * To obtain such a permit, you should contact New Cloud Technologies, Ltd.
 * at http://ncloudtech.com/contact.html
 *
 */

#include "aggregate.h"

using namespace cpp_integration;

Aggregate::Aggregate(Point* point, Array<String*>* array, String* s, Number* n)
{
    set("point", point);
    set("array", array);
    set("s", s);
    set("n", n);
}

Point* Aggregate::getPoint() const
{
    return get<Point*>("point");
}
Array<String*>* Aggregate::getStringArray() const
{
    return get<Array<String*>*>("array");
}
String* Aggregate::getString() const
{
    return get<String*>("s");
}
Number* Aggregate::getNumber() const
{
    return get<Number*>("n");
}

PointPair::PointPair(Number* x1, Number* y1, Number* x2, Number* y2)
    : topLeft(new Point{x1, y1})
    , bottomRight(new Point{x2, y2})
{
}

Point* PointPair::getTopLeft() const
{
    return topLeft;
}

Point* PointPair::getBottomRight() const
{
    return bottomRight;
}

RectHolder::RectHolder(Rect* rect)
    : rect(rect)
{
}

Rect* RectHolder::getRect() const
{
    return rect;
}

LargerAggregate::LargerAggregate(Number* x1, Number* y1, Number* x2, Number* y2)
{
    auto pair = new PointPair(x1, y1, x2, y2);
    auto rect = new Rect(pair->getTopLeft(), pair->getBottomRight());
    auto holder = new RectHolder(rect);

    set("pointPair", pair);
    set("rectHolder", holder);
}

PointPair* LargerAggregate::pointPair() const
{
    return get<PointPair*>("pointPair");
}

RectHolder* LargerAggregate::rectHolder() const
{
    return get<RectHolder*>("rectHolder");
}

Point* LargerAggregate::getTopLeft() const
{
    return pointPair()->getTopLeft();
}

Point* LargerAggregate::getBottomRight() const
{
    return pointPair()->getBottomRight();
}

Rect* LargerAggregate::getRect() const
{
    return rectHolder()->getRect();
}

LargerAggregate* LargerAggregate::getScaled(Number* factor) const
{
    auto tl = getTopLeft();
    auto br = getBottomRight();

    return new LargerAggregate{tl->x()->mul(factor), tl->y()->mul(factor), br->x()->mul(factor), br->y()->mul(factor)};
}